app/routes/user/password.tsx

```diff
@@ -0,0 +1,223 @@
+
+  invariant(oldPassword && newPassword)
+  return await updatePassword(request, oldPassword, newPassword)
+}
+
+export function WrongIdp() {
+  return (
+    <Alert type="error" heading="Error:" headingLevel="h4">
+      You must be signed in with a username and password to reset your password.
+    </Alert>
+  )
+}
+
+export function ResetPassword() {
+  const fetcher = useFetcher<typeof action>()
+  const errorMessage = fetcher.data
+
+  const [oldPassword, setOldPassword] = useState('')
+  const [newPassword, setNewPassword] = useState('')
+  const [isNewPasswordTouched, setIsNewPasswordTouched] = useState(false)
+  const [visible, setVisible] = useState(false)
+  const [interacted, setInteracted] = useState(false)
+
+  const containsLower = /[a-z]/.test(newPassword)
+  const containsUpper = /[A-Z]/.test(newPassword)
+  const containsNumber = /[0-9]/.test(newPassword)
+  const validLength = newPassword.length >= 8
+  const whitespaceValid = !/^\s|\s$/.test(newPassword)
+  const containsSpecialChar = /[~`!@#$%^.&*+=\-_ [\]\\';,/{}()|\\":<>?]/g.test(
+    newPassword
+  )
+
+  const valid =
+    whitespaceValid &&
+    containsLower &&
+    containsUpper &&
+    containsNumber &&
+    validLength &&
+    containsSpecialChar
+  const shouldDisableSubmit = !valid || !oldPassword
+
+  const formRef = useRef<HTMLFormElement>(null)
+
+  useEffect(() => {
+    if (fetcher.state === 'submitting') {
+      setInteracted(false)
+    }
+    if (fetcher.state === 'idle' && fetcher.data === null && formRef.current) {
+      formRef.current.reset()
+      setInteracted(false)
+    } else if (formRef && formRef.current) {
+      // this works around autofill on the old password field for all browsers
+      setOldPassword((formRef.current[0] as HTMLInputElement).value || '')
+    }
+  }, [fetcher.state, fetcher.data])
+
+  return (
+    <>
+      <h1>Reset Password</h1>
+      <>
+        <fetcher.Form method="POST" ref={formRef}>
+          <Label htmlFor="oldPassword">Old Password</Label>
+          <TextInput
+            data-focus
+            className={
+              errorMessage && !interacted
+                ? 'usa-input--error'
+                : 'margin-bottom-4'
+            }
+            name="oldPassword"
+            id="oldPassword"
+            type="password"
+            placeholder="Old Password"
+            onChange={(e) => {
+              setOldPassword(e.target.value)
+              setInteracted(true)
+            }}
+          />
+          {errorMessage === 'NotAuthorizedException' && !interacted && (
+            <div className="text-red margin-bottom-105 margin-top-2px">
+              Invalid Password
+            </div>
+          )}
+          {errorMessage === 'LimitExceededException' && !interacted && (
+            <div className="text-red margin-bottom-105 margin-top-2px">
+              Attempts Exceeded. Please try again later.
+            </div>
+          )}
```

Comment:
- [ ] Remember to reserve space for the message, so that the display doesn't bounce around: if neither message is displayed, then you can show a `<div>` containing nothing but `&nbsp;`. Also, you can lift the style `<div>` outside to eliminate duplicate code.
- [ ] Clear the error _when the user starts changing the old password field_, not when the user interacts with _any_ field in the form.

Reply:
@Courey, do you have sufficient access in GitHub to tick off the checklist items in the above comment?